src/compiler/checker.ts

```diff
@@ -4735,6 +4859,1157 @@ namespace ts {
+                    !getDeclarationWithTypeAnnotation(hostSymbol) &&
+                    !(typeToSerialize.symbol && some(typeToSerialize.symbol.declarations, d => getSourceFileOfNode(d) !== ctxSrc)) &&
+                    !some(getPropertiesOfType(typeToSerialize), p => isLateBoundName(p.escapedName)) &&
+                    !some(getPropertiesOfType(typeToSerialize), p => some(p.declarations, d => getSourceFileOfNode(d) !== ctxSrc)) &&
+                    every(getPropertiesOfType(typeToSerialize), p => isIdentifierText(symbolName(p), languageVersion) && !isStringAKeyword(symbolName(p)));
+                }
+
+                function makeSerializePropertySymbol<T extends Node>(createProperty: (
+                    decorators: readonly Decorator[] | undefined,
+                    modifiers: readonly Modifier[] | undefined,
+                    name: string | PropertyName,
+                    questionOrExclamationToken: QuestionToken | undefined,
+                    type: TypeNode | undefined,
+                    initializer: Expression | undefined
+                ) => T, methodKind: SyntaxKind): (p: Symbol, isStatic: boolean, baseType: Type | undefined) => (T | T[]) {
+                    return function serializePropertySymbol(p: Symbol, isStatic: boolean, baseType: Type | undefined) {
+                        if (isStatic && (p.flags & (SymbolFlags.Type | SymbolFlags.Namespace | SymbolFlags.Alias))) {
+                            // Only value-only-meaning symbols can be correctly encoded as class statics, type/namespace/alias meaning symbols
+                            // need to be merged namespace members
+                            return [];
+                        }
+                        if (p.flags & SymbolFlags.Prototype || (baseType && getPropertyOfType(baseType, p.escapedName)
+                            && isReadonlySymbol(getPropertyOfType(baseType, p.escapedName)!) === isReadonlySymbol(p)
+                            && (p.flags & SymbolFlags.Optional) === (getPropertyOfType(baseType, p.escapedName)!.flags & SymbolFlags.Optional)
+                            && isTypeIdenticalTo(getTypeOfSymbol(p), getTypeOfPropertyOfType(baseType, p.escapedName)!))) {
+                            return [];
+                        }
+                        const staticFlag = isStatic ? ModifierFlags.Static : 0;
+                        const rawName = unescapeLeadingUnderscores(p.escapedName);
+                        const name = getPropertyNameNodeForSymbolFromNameType(p, context) || createIdentifier(rawName);
+                        if (p.flags & (SymbolFlags.Property | SymbolFlags.Accessor | SymbolFlags.Variable)) {
+                            return setTextRange(createProperty(
+                                /*decorators*/ undefined,
+                                createModifiersFromModifierFlags((isReadonlySymbol(p) ? ModifierFlags.Readonly : 0) | staticFlag),
+                                name,
+   
```

Comment:
1. It reuses the existing node which means it preserves input type structure, alias usage, and comments. It's practically required for reasonable, as without it we go structural _almost all the time_, or even worse, fail to find a way to name something (because of some nested generativity that exposes a type from elsewhere when simplified).

2. Yes. (See above)

3. No? The important bits here are threefold - mapping jsdoc type nodes into their TS equivalents (yes, something in one of the quickfixes does that by parsing and reemitting via the node builder, but the point here is to _not use the node builder_, so we get better-than-node-builder print back), rewriting import types to use emitted module specifiers (nothing else does this, other than some as-you-descend code in the declaration emitter), and lastly, checking the visibility of all references we find as we go thru the tree (also something the declaration emitter does as-it-descends thru existing nodes).

Reply:
1. I didn't realise we relied on existing nodes so much. Interesting.
3. I was thinking of the quickfix code, I think.